[1.15] Vertical View Distance

AidanLovelace opened the pull request:
This new PR was created because I did some git nonos on the other branch so I recreated the changes on this branch.

Issue: #519

AidanLovelace committed:
Made it build on MacOS
(In addition to not having an entry for the natives for MacOS) It didn't previously because java on Mac allows a more recently loaded classpath to override another. This was problematic here because Minecraft 1.15.2 uses LWJGL 3.2.1 while we were using LWJGL 3.2.2, though that doesn't fully explain it. I suspect it had something to do with the natives overriding and not matching the version of Minecraft natives while minecraft was still using the 3.2.1 API. TL;DR I decremented the incremental version number and it fixed it. :D

AidanLovelace committed:
Added forge config
- Vertical view distance config option affects the view frustum

AidanLovelace committed:
Vertical view distance slider in video settings
- Properly updates renderer

AidanLovelace committed:
Vertical view distance!!!!
Very hacky. Not done.

AidanLovelace committed:
Fixed some oversights with the PlayerCubeTicketTracker graphs

AidanLovelace committed:
Some refactoring

NotStirred commented:
This is not working correctly, it seems to continuously load cubes around the player. Even at render distance 2h2v the same thing happens

AidanLovelace committed:
Fixed accidental infinite view distance
 - We now adjust the vertical view distance the same way normal view distance is
 - Fixed a potential bug in CubeDistanceGraph where propagation may not properly be limited by your range restrictions
 - Fixed incorrect arguments passed for isDecreasing within the directional graphs
 - Stopped doubling the horizontal view distance
 - Properly subtracting 2 from the view distances when check whether a cube is in range

AidanLovelace commented:
I think I got it properly working now. There were even more oversights. Guess I needed a week not looking at code lol.

NotStirred commented on src/main/java/io/github/opencubicchunks/cubicchunks/CubicChunksConfig.java:
```diff
@@ -0,0 +1,30 @@
+    static {
+        verticalViewDistance = builder
+                .comment("Vertical radius of cubes to load")
+                .defineInRange("verticalViewDistance", 4, 2, 16);
```
Match vanilla view distance range
```suggestion
                .defineInRange("verticalViewDistance", 4, 2, 32);
```

NotStirred commented on src/main/java/io/github/opencubicchunks/cubicchunks/chunk/ticket/PlayerCubeTicketTracker.java:
```diff
@@ -11,81 +12,127 @@
-                    iTicketManager.getCubeTaskPriorityQueueSorter().onUpdateCubeLevel(CubePos.from(i), () -> this.distances.get(i), k, (ix) -> {
-                        if (ix >= this.distances.defaultReturnValue()) {
-                            this.distances.remove(i);
+                    iTicketManager.getCubeTaskPriorityQueueSorter().onUpdateCubeLevel(CubePos.from(pos), () -> this.horizDistances.get(pos), horizCurrentDistance, (horizDistance) -> {
```
Take vertical view distance into account when sorting cube load tasks
```suggestion
                    iTicketManager.getCubeTaskPriorityQueueSorter().onUpdateCubeLevel(CubePos.from(pos), () -> this.horizDistances.get(pos), horizCurrentDistance+verticalViewDistance, (horizDistance) -> {
```

NotStirred reviewed (changes requested):
I assume you've rebased to most recent commit, so I haven't addressed those changes.

With a view distance of 8h8v the vertical view is 2 larger than horiz as shown by the vis, so something is not quite right there

AidanLovelace committed:
Made vertical view distance match the vanilla view distance range

Co-authored-by: Tom Martin <notstirred9@gmail.com>

AidanLovelace committed:
Take vertical view distance into account when sorting cube load tasks

Co-authored-by: Tom Martin <notstirred9@gmail.com>

AidanLovelace committed:
Fixed vertical view distance off-by-two mistake.

AidanLovelace committed:
Merge branch 'MC_1.15' into MC_1.15-VertViewDistance

NotStirred commented on build.gradle:
```diff
@@ -132,7 +132,7 @@ repositories{
 
 // for vulkan
-project.ext.lwjglVersion = "3.2.2"
+project.ext.lwjglVersion = "3.2.1"
```
```suggestion
project.ext.lwjglVersion = "3.2.2"
```

NotStirred commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinVideoSettingsScreen.java:
```diff
@@ -0,0 +1,39 @@
+
+    private static final AbstractOption[] MODIFIED_OPTIONS;
+
+    private static final SliderPercentageOption VERTICAL_RENDER_DISTANCE = new SliderPercentageOption("options.renderDistance", 2.0D, 16.0D, 1.0F, (gameSettings) -> (double) CubicChunksConfig.verticalViewDistance.get(),
```
```suggestion
    private static final SliderPercentageOption VERTICAL_RENDER_DISTANCE = new SliderPercentageOption("options.renderDistance", 2.0D, 32.0D, 1.0F, (gameSettings) -> (double) CubicChunksConfig.verticalViewDistance.get(),
```

NotStirred reviewed (commented):
Having issues with changing view distance not updating.
Also vertical view of 2 seems to break everything.

Barteks2x commented:
Merged manually, finished and fixed in https://github.com/OpenCubicChunks/CubicChunks/tree/MC_1.17
